### PR TITLE
[SPARK-38581][PYTHON][DOCS] List of supported pandas APIs for pandas-on-Spark docs.

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

This PR proposes to add new page named "Supported pandas APIs" for pandas-on-Spark documents.

### Why are the changes needed?

To let users can more easily find out whether a specific pandas API and its parameters are supported or not from the single document page.

### Does this PR introduce _any_ user-facing change?

Yes, the "Supported pandas APIs" page is added to the user guide for pandas API on Spark documents as below:

Screen Shot 2022-04-18 at 10 05 30 AM

### How was this patch tested?

Manually check the links in the documents & the existing doc build should be passed.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
@itholic can you take a look at the build failure?

```
/__w/spark/spark/python/docs/source/user_guide/pandas_on_spark/supported_pandas_api.rst:527:Inline literal start-string without end-string.
```

File: python/docs/source/user_guide/pandas_on_spark/supported_pandas_api.rst [hunk 1]
```diff
@@ -0,0 +1,803 @@
...
+
+In addition, the index will be lost if the ``index_col`` parameter is not specified for ``DataFrame.to_*`` APIs.
+
+====================== ======================= ====================================================================================================================================================================================================================================================================================================================================================================================
+API                    Implemented             Missing parameters
+====================== ======================= ====================================================================================================================================================================================================================================================================================================================================================================================
```
Comment (HyukjinKwon) [inline python/docs/source/user_guide/pandas_on_spark/supported_pandas_api.rst:43]: @itholic can we add some newlines into the data cells, and make this bars less? Seems ugly in the test format

File: python/docs/source/user_guide/pandas_on_spark/supported_pandas_api.rst [hunk 2]
```diff
@@ -0,0 +1,803 @@
...
+
+If there is non-implemented pandas API or parameter you want, you can create an `Apache Spark JIRA <https://issues.apache.org/jira/projects/SPARK/summary>`_ to request or to contribute by your own.
+
+The API list is updated based on the `latest pandas official API reference <https://pandas.pydata.org/docs/reference/index.html#>`_.
+
+
+Supported I/O APIs
+------------------
+
+In general, the ``read_*`` APIs can be very expensive if the ``index_col`` parameter is not specified since it attaches the `default index <https://spark.apache.org/docs/latest/api/python/user_guide/pandas_on_spark/options.html#default-index-type>`_.
```
Comment (HyukjinKwon) [inline python/docs/source/user_guide/pandas_on_spark/supported_pandas_api.rst:37]: Let's keep the lines within 100 length.

**Comment by xinrong-databricks** [comment]
Thanks @itholic , that's super helpful! Out of curiosity, did you check missing parameters manually?

**Comment by xinrong-databricks** [comment]
How do you think adding a comment in the [missing](https://github.com/apache/spark/tree/master/python/pyspark/pandas/missing) folder to remind people of managing the supported_pandas_api doc page?

**Comment by itholic** [comment]
> Thanks @itholic , that's super helpful! Out of curiosity, did you check missing parameters manually?

Yeah, I manually checked 😹 

> How do you think adding a comment in the [missing](https://github.com/apache/spark/tree/master/python/pyspark/pandas/missing) folder to remind people of managing the supported_pandas_api doc page?

Sounds good!! Let me add

File: python/docs/source/user_guide/pandas_on_spark/supported_pandas_api.rst (hunk 1 above)
Comment (itholic) [reply python/docs/source/user_guide/pandas_on_spark/supported_pandas_api.rst]: Just reformed!

**Comment by itholic** [comment]
Seems like the test is passed in my forked repo: https://github.com/itholic/spark/actions/runs/2158776052

File: python/docs/source/user_guide/pandas_on_spark/supported_pandas_api.rst [hunk 3]
```diff
@@ -0,0 +1,1320 @@
...
++-----------------------+----------+--------------------------------------------------------------+
+| idxmin                | O        | ``axis``                                                     |
++-----------------------+----------+--------------------------------------------------------------+
+| iloc                  | O        |                                                              |
++-----------------------+----------+--------------------------------------------------------------+
+| index                 | O        |                                                              |
++-----------------------+----------+--------------------------------------------------------------+
+| infer_objects         | X        |                                                              |
++-----------------------+----------+--------------------------------------------------------------+
+| interpolate           | X        |                                                              |
```
Comment (HyukjinKwon) [inline python/docs/source/user_guide/pandas_on_spark/supported_pandas_api.rst:354]: Let's keep the PR up to date. e.g.) this is implemented https://github.com/apache/spark/commit/30fc0ba2307dcb71de1ebf785b96b4e5f05d96bd

File: python/docs/source/user_guide/pandas_on_spark/supported_pandas_api.rst [hunk 4]
```diff
@@ -0,0 +1,1320 @@
...
++----------------------+----------+---------------------------------------------------------------+
+| read_stata           | X        |                                                               |
++----------------------+----------+---------------------------------------------------------------+
+| DataFrame.to_stata   | X        |                                                               |
++----------------------+----------+---------------------------------------------------------------+
+
+Supported General Function APIs
+-------------------------------
+
+=============== =========== ===============================================================
```
Comment (HyukjinKwon) [inline python/docs/source/user_guide/pandas_on_spark/supported_pandas_api.rst:146]: Can we have the same style of chart?

File: python/docs/source/user_guide/pandas_on_spark/supported_pandas_api.rst [hunk 5]
```diff
@@ -0,0 +1,1320 @@
...
+| var                   | O        | ``skipna``, ``level``                                        |
++-----------------------+----------+--------------------------------------------------------------+
+| view                  | X        |                                                              |
++-----------------------+----------+--------------------------------------------------------------+
+| where                 | O        | ``inplace``, ``axis``, ``level``, ``errors``                 |
++-----------------------+----------+--------------------------------------------------------------+
+| xs                    | O        | ``axis``, ``drop_level``                                     |
++-----------------------+----------+--------------------------------------------------------------+
+
+Supported DataFrame APIs
```
Comment (HyukjinKwon) [inline python/docs/source/user_guide/pandas_on_spark/supported_pandas_api.rst:620]: Let's place DataFrame APIs on the top.

File: python/docs/source/user_guide/pandas_on_spark/supported_pandas_api.rst [hunk 6]
```diff
@@ -0,0 +1,1320 @@
...
++------------------+-----------+-------------------------------------------------------------------+
+| empty            | O         |                                                                   |
++------------------+-----------+-------------------------------------------------------------------+
+| eq               | O         | ``axis``, ``level``                                               |
++------------------+-----------+-------------------------------------------------------------------+
+| equals           | O         | ``axis``, ``level``                                               |
++------------------+-----------+-------------------------------------------------------------------+
+| eval             | O         |                                                                   |
++------------------+-----------+-------------------------------------------------------------------+
+| ewm              | X         |                                                                   |
```
Comment (HyukjinKwon) [inline python/docs/source/user_guide/pandas_on_spark/supported_pandas_api.rst:739]: This is implemented too.

File: python/docs/source/user_guide/pandas_on_spark/supported_pandas_api.rst [hunk 7]
```diff
@@ -0,0 +1,1320 @@
...
+JIRA <https://issues.apache.org/jira/projects/SPARK/summary>`__ to request or to contribute by your
+own.
+
+The API list is updated based on the `latest pandas official API
+reference <https://pandas.pydata.org/docs/reference/index.html#>`__.
+
+Supported I/O APIs
+------------------
+
+In general, the ``read_*`` APIs can be very expensive if the ``index_col`` parameter is not
```
Comment (HyukjinKwon) [inline python/docs/source/user_guide/pandas_on_spark/supported_pandas_api.rst:41]: Let's remove this

**Review by HyukjinKwon — COMMENTED** [review]
Looks pretty good otherwise.

File: python/docs/source/user_guide/pandas_on_spark/supported_pandas_api.rst (hunk 3 above)
Comment (itholic) [reply python/docs/source/user_guide/pandas_on_spark/supported_pandas_api.rst:354]: Seems like `interpolate` and `ewm` will be supported from 3.4.0.

Maybe can we separate the PR for 3.3.0 and master ?

Comment (HyukjinKwon) [reply python/docs/source/user_guide/pandas_on_spark/supported_pandas_api.rst:354]: Yeah, let's fix it here, and create another PR to remove it out for branch-3.3.

**Comment by HyukjinKwon** [comment]
Build: https://github.com/itholic/spark/runs/6001521275

File: python/docs/source/user_guide/pandas_on_spark/supported_pandas_api.rst (hunk 3 above)
Comment (itholic) [reply python/docs/source/user_guide/pandas_on_spark/supported_pandas_api.rst:354]: Sounds good

**Comment by itholic** [comment]
Build: https://github.com/itholic/spark/actions/runs/2181653453

**Comment by itholic** [comment]
Build: https://github.com/itholic/spark/actions/runs/2184725527

**Comment by xinrong-databricks** [comment]
LGTM, thanks!

**Comment by ueshin** [comment]
I'm wondering whether `O` and `X` globally mean Yes/True and No/False.
I'd suggest we just use `Y` and `N` instead.
Additionally, I'd like to add `P` for partially implemented, meaning some parameters may not be supported yet or something.

**Comment by HyukjinKwon** [comment]
yeah, I agree with ^.

**Comment by itholic** [comment]
Good idea. Just addressed!

**Comment by itholic** [comment]
Build: https://github.com/itholic/spark/actions/runs/2187151624

**Comment by Yikun** [comment]
I guess we might want to consider below ideas:

1. Generate these automically and more accurate. See end script it's a demo to show how to find diff in Series, I believe we can easy to expanda this script to all objects. (from results we can see we lack some diff in current PR.)
```python
import pandas as pd
import pyspark.pandas as ps
from inspect import getmembers, isfunction

pds = pd.Series
pds_funcs = dict([m for m in getmembers(pds, isfunction) if not m[0].startswith("_")])
pss = ps.Series
pss_funcs = dict([m for m in getmembers(pss, isfunction) if not m[0].startswith("_")])

# See what function in pandas but not in spark
pds_funcs.keys() - pss_funcs.keys()

supported_func = pds_funcs.keys() & pss_funcs.keys()
for name in supported_func:
    pds_func = pds_funcs[name]
    pss_func = pss_funcs[name]
    # See what func parameter supported in pandas but not supported in pyspark.pandas
    diff = set(inspect.signature(pds_funcs[name]).parameters) - set(inspect.signature(pss_funcs[name]).parameters)
    print(name, diff)
```


<details>
  <summary>Results</summary>
Unsuported method (N):

```
['asfreq', 'combine', 'convert_dtypes', 'infer_objects', 'info', 'memory_usage', 'ravel', 'reorder_levels', 'resample', 'searchsorted', 'set_axis', 'set_flags', 'slice_shift', 'to_hdf', 'to_period', 'to_pickle', 'to_sql', 'to_timestamp', 'to_xarray', 'tshift', 'tz_convert', 'tz_localize', 'view']
```

Unsupported parameters (P):

```
abs []
add ['axis', 'fill_value', 'level']
add_prefix []
add_suffix []
agg ['args', 'axis', 'kwargs']
aggregate ['args', 'axis', 'kwargs']
align ['broadcast_axis', 'fill_axis', 'fill_value', 'level', 'limit', 'method']
all ['bool_only', 'kwargs', 'level']
any ['bool_only', 'kwargs', 'level', 'skipna']
append []
apply ['convert_dtype', 'kwargs']
argmax ['args', 'axis', 'kwargs', 'skipna']
argmin ['args', 'axis', 'kwargs', 'skipna']
argsort ['axis', 'kind', 'order']
asof ['subset']
astype ['copy', 'errors']
at_time []
autocorr ['lag']
backfill ['downcast']
between []
between_time ['inclusive']
bfill ['downcast']
bool []
clip ['args', 'axis', 'kwargs']
combine_first []
compare ['align_axis']
copy []
corr ['min_periods']
count ['level']
cov ['ddof']
cummax ['args', 'axis', 'kwargs']
cummin ['args', 'axis', 'kwargs']
cumprod ['args', 'axis', 'kwargs']
cumsum ['args', 'axis', 'kwargs']
describe ['datetime_is_numeric', 'exclude', 'include']
diff []
div ['axis', 'fill_value', 'level']
divide ['axis', 'fill_value', 'level']
divmod ['axis', 'fill_value', 'level']
dot []
drop ['axis', 'columns', 'errors', 'inplace']
drop_duplicates []
droplevel ['axis']
dropna ['how']
duplicated []
eq ['axis', 'fill_value', 'level']
equals []
ewm ['adjust', 'axis', 'ignore_na', 'method', 'times']
expanding ['axis', 'center', 'method']
explode ['ignore_index']
factorize []
ffill ['downcast']
fillna ['downcast']
filter []
first []
first_valid_index []
floordiv ['axis', 'fill_value', 'level']
ge ['axis', 'fill_value', 'level']
get []
groupby ['group_keys', 'level', 'observed', 'sort', 'squeeze']
gt ['axis', 'fill_value', 'level']
head []
hist ['ax', 'backend', 'by', 'figsize', 'grid', 'kwargs', 'legend', 'xlabelsize', 'xrot', 'ylabelsize', 'yrot']
idxmax ['args', 'axis', 'kwargs']
idxmin ['args', 'axis', 'kwargs']
interpolate ['axis', 'downcast', 'inplace', 'kwargs', 'limit_area', 'limit_direction']
isin []
isna []
isnull []
item []
items []
iteritems []
keys []
kurt ['kwargs', 'level', 'skipna']
kurtosis ['kwargs', 'level', 'skipna']
last []
last_valid_index []
le ['axis', 'fill_value', 'level']
lt ['axis', 'fill_value', 'level']
mad ['axis', 'level', 'skipna']
map []
mask ['axis', 'errors', 'inplace', 'level', 'try_cast']
max ['kwargs', 'level', 'skipna']
mean ['kwargs', 'level', 'skipna']
median ['kwargs', 'level', 'skipna']
min ['kwargs', 'level', 'skipna']
mod ['axis', 'fill_value', 'level']
mode []
mul ['axis', 'fill_value', 'level']
multiply ['axis', 'fill_value', 'level']
ne ['axis', 'fill_value', 'level']
nlargest ['keep']
notna []
notnull []
nsmallest ['keep']
nunique []
pad ['downcast']
pct_change ['fill_method', 'freq', 'kwargs', 'limit']
pipe []
pop []
pow ['axis', 'fill_value', 'level']
prod ['kwargs', 'level', 'skipna']
product ['kwargs', 'level', 'skipna']
quantile ['interpolation']
radd ['axis', 'fill_value', 'level']
rank ['axis', 'na_option', 'pct']
rdiv ['axis', 'fill_value', 'level']
rdivmod ['axis', 'fill_value', 'level']
reindex ['args', 'kwargs']
reindex_like ['copy', 'limit', 'method', 'tolerance']
rename ['axis', 'copy', 'errors', 'inplace', 'level']
rename_axis ['axis', 'columns', 'copy']
repeat ['axis']
replace ['inplace', 'limit', 'method']
reset_index []
rfloordiv ['axis', 'fill_value', 'level']
rmod ['axis', 'fill_value', 'level']
rmul ['axis', 'fill_value', 'level']
rolling ['axis', 'center', 'closed', 'method', 'on', 'win_type']
round ['args', 'kwargs']
rpow ['axis', 'fill_value', 'level']
rsub ['axis', 'fill_value', 'level']
rtruediv ['axis', 'fill_value', 'level']
sample ['axis', 'ignore_index', 'weights']
sem ['kwargs', 'level', 'skipna']
shift ['axis', 'freq']
skew ['kwargs', 'level', 'skipna']
sort_index ['ignore_index', 'key', 'sort_remaining']
sort_values ['axis', 'ignore_index', 'key', 'kind']
squeeze []
std ['kwargs', 'level', 'skipna']
sub ['axis', 'fill_value', 'level']
subtract ['axis', 'fill_value', 'level']
sum ['kwargs', 'level', 'skipna']
swapaxes ['axis1', 'axis2']
swaplevel []
tail []
take ['axis', 'is_copy', 'kwargs']
to_clipboard []
to_csv ['chunksize', 'compression', 'decimal', 'doublequote', 'encoding', 'errors', 'float_format', 'index', 'index_label', 'line_terminator', 'path_or_buf', 'quoting', 'storage_options']
to_dict []
to_excel ['storage_options']
to_frame []
to_json ['date_format', 'date_unit', 'default_handler', 'double_precision', 'force_ascii', 'indent', 'index', 'path_or_buf', 'storage_options']
to_latex ['caption', 'label', 'position']
to_list []
to_markdown ['index', 'kwargs', 'storage_options']
to_numpy ['copy', 'dtype', 'kwargs', 'na_value']
to_string ['min_rows']
tolist []
transform []
transpose []
truediv ['axis', 'fill_value', 'level']
truncate []
unique []
unstack ['fill_value']
update []
value_counts []
var ['kwargs', 'level', 'skipna']
where ['axis', 'errors', 'inplace', 'level', 'try_cast']
xs ['axis', 'drop_level']
```

</details>

2. Should we also mentioned specific panda version? After this, we could easy to see diff between pandas and pyspark.pandas.

**Comment by itholic** [comment]
@Yikun Yeah, I tried it at first, but it was a bit tricky since the rules for the RST format that Sphinx checks are way stricter than I think when building a document, so I decided to manually write the initial version first on this PR.

Let me create an another ticket, and can we improve it there ??

**Comment by Yikun** [comment]
@itholic Fine for me, and actually I wasn't mean use above scripts to generate doc completely (inline code), because we had some specific note based on doc.

You could just regard it as dev helper script now and help you making result more accurate. :)

**Review by Yikun — APPROVED** [review]
We can improve them later for sure

**Review by HyukjinKwon — APPROVED** [review]

**Review by ueshin — APPROVED** [review]

**Comment by itholic** [comment]
@Yikun created ticket for improving the doc generation here: SPARK-38961 FYI.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by HyukjinKwon** [comment]
@itholic mind creating a PR for branch-3.3?

**Comment by itholic** [comment]
@HyukjinKwon Just created at https://github.com/apache/spark/pull/36308 !

**Pull request closed without merging**